File: server/js-thesis-project/src/transactions/new_order_transaction.js
const {
  get_random_warehouse,
  get_random_district_from_warehouse,
  get_random_customer_from_warehouse_district,
  get_items,
  random_int
} = require('../utils');

async function do_new_order(models) {
  // Get Warehouse
  const warehouse = await get_random_warehouse(models)
  const w_id = warehouse.w_id

  // Get District
  const district = await get_random_district_from_warehouse(models, warehouse)
  const d_id = district.d_id

  // Get Customer
  const customer = await get_random_customer_from_warehouse_district(models, warehouse, district)

  // Update District
  const next_o_id = district.d_next_o_id + 1
  await models.district.update({d_next_o_id: next_o_id}, {
    where: {
      d_id: d_id,
      d_w_id: w_id
    }
  })

  // Create Order
  const order = await models.orders.create({
    o_id: next_o_id,
    o_d_id: d_id,
    o_w_id: w_id,
    o_c_id: customer.c_id,
  })

  // Create New Order
  await models.new_order.create({
    no_o_id: order.o_id,
    no_d_id: d_id,
    no_w_id: w_id
  })

  const items = await get_items(models)
  const items_length = items.length
  for (let idx = 0; idx < items_length; idx++) {
    const item = items[idx]

    // Get Stock
    const stocks = await models.stock.findAll({
      where: {
        s_i_id: item.i_id,
        s_w_id: w_id
      }
    })
    const stock = stocks[0]
    if (stock == null) {
      return;
    }

    // Update Stock
    const ol_quantity = random_int(1, 10)
    const s_ytd = stock.s_ytd + ol_quantity
    let s_quantity
    if (stock.s_quantity >= ol_quantity + 10) {
      s_quantity = stock.s_quantity - ol_quantity
    } else {
      s_quantity = stock.s_quantity + 91 - ol_quantity
    }
    const s_order_cnt = stock.s_order_cnt + 1
    await models.stock.update({
      s_ytd,
      s_quantity,
      s_order_cnt
    }, {
      where: {
        s_i_id: item.i_id,
        s_w_id: w_id
      }
    })

    // Create Order Line
    const now = new Date().toUTCString();
    await models.order_line.create({
      ol_o_id: order.o_id,
      ol_d_id: d_id,
      ol_w_id: w_id,
      ol_number: idx + 1,
      ol_i_id: stock.s_i_id,
      ol_supply_w_id: stock.s_w_id,
      ol_delivery_d: now,
      ol_quantity: ol_quantity,
      ol_amount: ol_quantity * item.i_price
    })
  }
}

module.exports = do_new_order;

File: server/python-thesis-project/api/transactions/new_order_transaction.py
import random

from django.utils import timezone

import api.utils as utils
from api.models import NewOrder, OrderLine, Orders, Stock


def do_new_order():
    # Get Warehouse
    warehouse = utils.get_random_warehouse()
    w_id = warehouse.w_id

    # Get District
    district = utils.get_random_district_from_warehouse(warehouse)
    d_id = district.d_id

    # Get Customer
    customer = utils.get_random_customer_from_warehouse_and_district(
        warehouse, district)

    # Update District
    district.d_next_o_id += 1
    district.save(update_fields=['d_next_o_id'])

    # Create Order
    order = Orders.objects.create(
        o_id=district.d_next_o_id,
        o_d_id=d_id,
        o_w_id=w_id,
        o_c=customer,
        o_entry_d=timezone.now(),
    )

    # Create New Order
    NewOrder.objects.create(
        no_o_id=order.o_id,
        no_d_id=d_id,
        no_w_id=w_id
    )

    # Get Item
    items = utils.get_items()
    for idx, item in enumerate(items):
        # Get Stock
        stock = Stock.objects.get(s_i_id=item.i_id, s_w__w_id=w_id)

        # Update Stock
        ol_quantity = random.randint(1, 10)
        stock.s_ytd += ol_quantity
        if stock.s_quantity >= ol_quantity + 10:
            stock.s_quantity -= ol_quantity
        else:
            stock.s_quantity += 91 - ol_quantity
        stock.s_order_cnt += 1
        stock.save(update_fields=['s_ytd', 's_quantity', 's_order_cnt'])

        # Create OrderLine
        OrderLine.objects.create(
            ol_o_id=order.o_id,
            ol_d_id=d_id,
            ol_w_id=w_id,
            ol_number=idx+1,
            ol_i=stock,
            ol_supply_w_id=stock.s_w.w_id,
            ol_delivery_d=timezone.now(),
            ol_quantity=ol_quantity,
            ol_amount=ol_quantity*item.i_price
        )
